### DISCUSSION TITLE
C-c C-c causes nvim-cmp havoc while q works fine

### DISCUSSION THREAD

**Discussion by r-darwish** [NeogitOrg/neogit#1492] [Q&A] [answered]
![image](https://github.com/user-attachments/assets/93a23818-a719-4989-a801-1b0710ceda07)

Whenever I commit using `q` in normal mode everything works fine. However, pressing `C-c C-c` in insert node causes tons of error messages and makes neovim unusable.

I suppose this isn't necessarly a bug in either extension but maybe misconfiguration. What is the difference between `q` and `C-c C-c` and what can I do to mitigate this error?

Using LazyVim with Neogit enabled.

Putting this line here to make the discussion search friendly. I couldn't copy the entire traceback as nvim is completley broken in this phase.
`...hare/nvim/lazy/nvim-cmp/lua/cmp/view/ghost_text_view.lua`

**Comment by CKolkey** [comment, marked as answer]
Let's have a look.

Here's the function run when with `<c-c><c-c>`: https://github.com/NeogitOrg/neogit/blob/master/lua/neogit/buffers/editor/init.lua#L202-L212
```lua
        [mapping_I["Submit"]] = function(buffer)
          logger.debug("[EDITOR] Action I: Submit")
          vim.cmd.stopinsert()
          if amend_header then
            buffer:set_lines(0, 0, false, amend_header)
            amend_header = nil
          end

          buffer:write()
          buffer:close(true)
        end,
```

Stops insert, writes the amend header to the file (if present), writes the buffer, and closes it with force (the `true`)

We can compare that to the normal-mode `q` function: https://github.com/NeogitOrg/neogit/blob/master/lua/neogit/buffers/editor/init.lua#L222-L235
```lua
        [mapping["Close"]] = function(buffer)
          logger.debug("[EDITOR] Action N: Close")
          if amend_header then
            buffer:set_lines(0, 0, false, amend_header)
            amend_header = nil
          end

          if buffer:get_option("modified") and not input.get_confirmation("Save changes?") then
            aborted = true
          end

          buffer:write()
          buffer:close(true)
        end,
```

Pretty much the same, except it has slightly different logic around aborting the commit.

Both of these would then trigger the `on_detach` callback here: https://github.com/NeogitOrg/neogit/blob/master/lua/neogit/buffers/editor/init.lua#L99-L117


My _guess_ here is some kind of race condition, where the `on_win` function in cmp (https://github.com/hrsh7th/nvim-cmp/blob/main/lua/cmp/view/ghost_text_view.lua#L35) is being invoked just after the buffer is removed, but before the decoration provider is stopped.

I suspect if you added a bit like this to the decoration provider function it would work:

```lua
    on_win = function(_, win)
      if self.extmark_buf and not vim.api.nvim_buf_is_loaded(self.extmark_buf) then
        return false
      end

      -- ... etc
```

**Reply by r-darwish** [reply]
Thank you. It seems to fix the problem. I opened [This PR](https://github.com/hrsh7th/nvim-cmp/pull/2053). Thank you for taking the time to help me debug another plugin.
